patroni/postgresql.py

```diff
@@ -165,49 +113,85 @@ def data_directory_empty(self):
+            
+        return mconn
 
     def create_replica(self, master_connection, env):
-        self.set_state('building replica from {host}:{port}'.format(**master_connection))
+        # create the replica according to the replica_method
+        # defined by the user.  this is a list, so we need to
+        # loop through all methods the user supplies
         connstring = self.build_connstring(master_connection)
-        cmd = self.config['restore']
-        try:
-            ret = subprocess.call(shlex.split(cmd) + [self.scope, "replica", self.data_dir, connstring], env=env)
-            self.delete_trigger_file()
-        except:
-            logger.exception('Error when creating replica')
-            ret = 1
-        if ret != 0:
-            self.set_state('failed to build replica from {host}:{port}'.format(**master_connection))
-        return ret
+        env = os.environ.copy()
+        env['PGPASSFILE'] = 'pgpass'
+        # get list of replica methods from config
+        replica_list = self.config.get('create_replica_method', 'basebackup')
+        replica_methods = [rm.strip() for rm in replica_list.split(',')]
+        # go through them in priority order
+        for replica_method in replica_methods:
+            # if the method is basebackup, then use the built-in
+            if replica_method == "basebackup":
+                ret = self.basebackup(connstring, env)
+                if ret == 0:
+                    # if basebackup succeeds, exit with success
+                    return 0
+            else:
+                # user-defined method; check for configuration
+                # not required, actually
+                if replica_method in self.config:
+                    # look to see if the user has supplied a full command path
+                    # if not, use the method name as the command
+                    if "command" in self.config[replica_method]:
+                        cmd = self.config[replica_method]["command"]
+                    else:
+                        cmd = replica_method
+                        
+                    # get the rest of the replica config
+                    method_config = self.config[replica_method].copy()
+                    # remove the command and turn it into a shlex set
+                    del method_config["command"]
+                    # add the default parameters
+                    method_config.update({"scope": self.scope, 
+                                          "role" : "replica",
+                                          "datadir" : self.data_dir,
+                                          "connstring" : self.connstring})
+                    params = ["--{0}={1}".format(arg, val) for arg, val in method_config.iteritems()]
+                    
+                try:
+                    # call script with the full set of parameters
+                    ret = subprocess.call(shlex.split(cmd) + shlex.split(method_config), env=env)
```

Comment:
I guess you wanted to call: subprocess.call(shlex.split(cmd) + **params**, env=env)

But! On this line we would have cmd and params undefined if there is no special configuration section for a given replica_method in the configuration file.